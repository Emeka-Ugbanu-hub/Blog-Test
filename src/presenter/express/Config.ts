
import Service from '../../service/Service';
import { LoggerInstance } from 'winston';

export default interface Config {
  readonly morganLogFormat: string,
  readonly morganDirectory: string,
  readonly service: Service,
  readonly logger: LoggerInstance;
}