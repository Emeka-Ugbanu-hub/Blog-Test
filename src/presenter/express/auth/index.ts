import login from './login';
import register from './register';
import forgetPassword from './forgetPassword';

export {
  login, 
  register,
  forgetPassword
}