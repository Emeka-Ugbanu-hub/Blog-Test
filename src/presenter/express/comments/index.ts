import createComment from './createComment';
import getCommentById from './getCommentById';
export {
  createComment,
  getCommentById
}