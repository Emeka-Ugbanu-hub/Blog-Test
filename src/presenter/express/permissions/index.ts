import createPermission from './createPermission';
import getPermissionById from './getPermissionById';
import getPermissions from './getPermissions';
import updatePermission from './updatePermission';
import deletePermissionById from './deletePermissionById';
export {
  getPermissions,
  getPermissionById,
  createPermission,
  updatePermission,
  deletePermissionById
}