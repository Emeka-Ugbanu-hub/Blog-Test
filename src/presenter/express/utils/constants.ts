const OK_200_HTTP_CODE = 200;
const CREATED_201_HTTP_CODE = 201;
const NO_CONTENT_204_HTTP_CODE = 204;
const BAD_REQUEST_400_HTTP_CODE = 400;
const UNAUTHORISED_401_HTTP_CODE = 401;
const FORBIDDEN_403_HTTP_CODE = 403;
const NOT_FOUND_404_HTTP_CODE = 404;
const CONFLICT_409_HTTP_CODE = 409;
const UNPROCESSABLE_ENTITY_422_HTTP_CODE = 422;
const SERVER_ERROR_500_HTTP_CODE = 500;


export {
  OK_200_HTTP_CODE,
  CREATED_201_HTTP_CODE,
  NO_CONTENT_204_HTTP_CODE,
  BAD_REQUEST_400_HTTP_CODE,
  UNAUTHORISED_401_HTTP_CODE,
  FORBIDDEN_403_HTTP_CODE,
  NOT_FOUND_404_HTTP_CODE,
  CONFLICT_409_HTTP_CODE,
  UNPROCESSABLE_ENTITY_422_HTTP_CODE,
  SERVER_ERROR_500_HTTP_CODE,
}