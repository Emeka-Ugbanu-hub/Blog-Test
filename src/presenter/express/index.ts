import { Router, Request, Response} from 'express';
import Config from './Config';
import enhancedRouter from './enhancedRouter';
import {register, login, forgetPassword,resetPassword} from './auth';
import {getPosts, createPost, getPostById, updatePost, deletePostById} from './posts';
import {getUserById, getUsers, updateUser, createUser, deleteUserById} from './users';
import {createComment, getCommentById, getComments, updateComment} from './comments';

export default (config: Config): Router => {
  const router: Router = enhancedRouter(config);

  router.get('/', (req: Request, res: Response) => {
    res.status(200)
       .json({message: "This is where the awesomeness happen..."});
  });

  router.post('/auth/register', register(config));
  router.post('/auth/login', login(config));
  router.post('/auth/forget-password', forgetPassword(config));
  router.post('/auth/reset-password', resetPassword(config));

  router.post('/users', createUser(config));
  router.delete('/users/:user_id', deleteUserById(config));
  router.get('/users/:user_id', getUserById(config));
  router.get('/users', getUsers(config));
  router.patch('/users/:user_id', updateUser(config));
  
  router.get('/posts', getPosts(config));
  router.get('/posts/:post_id', getPostById(config));
  router.delete('/posts/:post_id', deletePostById(config));
  router.post('/posts', createPost(config));
  router.patch('/posts/:post_id', updatePost(config));

  router.get('/comments', getComments(config));
  router.post('/comments', createComment(config));
  router.get('/comments/:comment_id', getCommentById(config));
  router.patch('/comments/:comment_id', updateComment(config));

  return router;
}