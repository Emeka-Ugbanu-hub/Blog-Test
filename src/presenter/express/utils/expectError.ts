import {UNPROCESSABLE_ENTITY_422_HTTP_CODE} from './constants';

interface Response {
  status: number;
  body: any;
}

export default (response: Response, expectedStatusCode: number = UNPROCESSABLE_ENTITY_422_HTTP_CODE) => {
  expect(response.status).toBe(expectedStatusCode);
  expect(response.body).toMatchSnapshot();
}