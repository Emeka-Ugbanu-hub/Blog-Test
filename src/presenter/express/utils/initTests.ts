import config from '../../../config';
import * as sourceMapSupport from 'source-map-support';
sourceMapSupport.install();
import * as iconvLite from 'iconv-lite';
iconvLite.encodingExists('foo');
import * as express from 'express';
import * as createSupertest from 'supertest';
import logger from '../../../logger';
import presenter from '../index';
import serviceFactory from '../../../service/factory';
import {API_ROUTE_V1} from '../../../utils/constants';

const app: express.Application = express();

const service = serviceFactory();

const presenterFacade = presenter({
  morganLogFormat: config.express.morganLogFormat,
  morganDirectory: config.express.morganDirectory,
  service,
  logger
});

app.use(API_ROUTE_V1, presenterFacade);

const request = createSupertest(app);

export default () => {

  beforeAll(async() => {
    await service.migrate();
  });

  beforeEach(async() => {
    await service.migrate();
  });
  
  afterEach(async() => {
    await service.rollback();
  });

  afterAll(async() => {
    await service.clearService();
  });
  
  return { service, request };
};