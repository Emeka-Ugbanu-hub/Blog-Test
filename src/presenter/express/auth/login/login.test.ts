import initTests from '../../utils/initTests';
import { API_ROUTE_V1 } from '../../../../utils/constants';
import { Response } from 'express';
import { OK_200_HTTP_CODE } from '../../utils/constants';
import config from '../../../../config';
import { TEST_VALID_LOGIN_USER, TEST_VALID_REGISTER_USER,TEST_INVALID_EMAIL,TEST_VALID_PASSWORD, TEST_VALID_EMAIL, TEST_TOO_SHORT_PASSWORD, TEST_DIFFERENT_VALID_PASSWORD} from '../../../../utils/testValues';
import * as R  from 'ramda';
import expectError from '../../utils/expectError';

describe(__filename, () => {

  const { service, request } = initTests();

  it('should fail to log in user without input', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/login`);
    expectError(response);
  });

  it('should fail to log in user when email is invalid', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/login`)
                                  .send({
                                    email: TEST_INVALID_EMAIL, 
                                    password: TEST_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to log in a user without password', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/login`)
                                  .send({email: TEST_VALID_EMAIL});
    expectError(response);
  });

  it('should fail to log in when user does not exist', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/login`)
                                  .send(TEST_VALID_LOGIN_USER);
    expectError(response);
  });

  it('should fail to log in a user when password is invalid', async () => {
    
    const registeredUser = await service.register(TEST_VALID_LOGIN_USER);
    
    const response = await request.post(`${API_ROUTE_V1}/auth/login`)
                                  .send({
                                    email: TEST_VALID_LOGIN_USER.email, 
                                    password: TEST_DIFFERENT_VALID_PASSWORD,
                                  });
    expectError(response);
  });

  it('should succesfully log in a user', async () => {
    
    const registeredUser = await service.register(TEST_VALID_REGISTER_USER);

    const response = await request.post(`${API_ROUTE_V1}/auth/login`)
                                  .send({email: TEST_VALID_REGISTER_USER.email, password: TEST_VALID_REGISTER_USER.password});

    const {user, token} = response.body;
    expect(response.status).toBe(OK_200_HTTP_CODE);
    expect(user.email).toEqual(TEST_VALID_REGISTER_USER.email);
  });

});