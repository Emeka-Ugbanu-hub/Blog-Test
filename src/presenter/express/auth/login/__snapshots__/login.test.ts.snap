// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/login/login.test.ts should fail to log in a user when password is invalid 1`] = `
Object {
  "message": "Incorrect email or password",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/login/login.test.ts should fail to log in a user without password 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.password'",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/login/login.test.ts should fail to log in user when email is invalid 1`] = `
Object {
  "errors": Array [
    "Invalid email in 'user.email'. Received 'undefined'",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/login/login.test.ts should fail to log in user without input 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.email'",
    "Missing required value in 'user.password'",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/login/login.test.ts should fail to log in when user does not exist 1`] = `
Object {
  "message": "Incorrect email or password",
}
`;
